t/plugin/proxy-rewrite.t

```diff
@@ -1249,3 +1249,41 @@ version: nginx_var_does_not_exist
 x-real-ip: 127.0.0.1
 --- no_error_log
 [error]
+
+
+
+=== TEST 42: set route(rewrite uri based on ctx.var)
```

Comment:
only set route is not enough. we need to make a test

Reply:
Unfortunately, I think this is the end of my understanding of how to contribute to APISIX. I never heard of Lua before this project so I'm relying on the documentation and contributors of this project to help me understand what to do.

Can you point to other tests (or documentation) which does everything I need to do in my test?

Reply:
I write a full test case example, you can make a try.

^_^

```
=== TEST 42: set route(rewrite uri based on ctx.var)
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/routes/1',
                 ngx.HTTP_PUT,
                 [[{
                        "plugins": {
                            "proxy-rewrite": {
                                "uri": "/$arg_new_uri"
                            }
                        },
                        "upstream": {
                            "nodes": {
                                "127.0.0.1:1980": 1
                            },
                            "type": "roundrobin"
                        },
                        "uri": "/test"
                }]]
                )

            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]



=== TEST 43: hit route(upstream uri: should be /hello)
--- request
GET /test?new_uri=hello
--- response_body
hello world
--- no_error_log
[error]

```

Reply:
Thanks for the help. I'm still not sure how your tests cover anything more than what my test covers though.

Based on your tests it appears tests are executed sequentially and share a context between them. The first test creates a route with the `proxy-rewrite` plugin rewriting the uri to `$arg_new_uri`. This appears to be the same thing my test covers with just a different variable to subsitute.

Your second test then appears to continue work from the first test and now hits `/test?new_uri=hello` and expects `hello world` as a response. That doesn't make sense to me. Where does `hello world` come from and how does hitting `/test?new_uri=hello` verify that `proxy-rewrite` has worked correctly? There should be some way to set the value of `$arg_new_uri` and verify the resulting value of uri after `proxy-rewrite` is equiavlent. Also, if `$arg_new_uri` is really the value of `new_uri` that is a bit confusing. I wouldn't have guessed adding `arg_` makes a variable come from url get params.

Apologies if I'm making this more difficult than it needs to be. I've tried reading the [documention](https://github.com/apache/apisix/blob/master/doc/plugin-develop.md#write-test-case) but it doesn't really explain to me how testing works.

Reply:
'arg_****': argument name in the request line

nginx variable link: http://nginx.org/en/docs/http/ngx_http_core_module.html#var_arg_

`$arg_new_uri`, it means to fetch the value from query string which name is `new_uri`.

1.  the current request is `/test?new_uri=hello`, so the `$arg_new_uri` should equal `hello`
2. `"uri": "/$arg_new_uri"` -> `"uri": "/hello"`
3.  response body is 'hello world' if the request is `/hello`: https://github.com/apache/apisix/blob/master/t/lib/server.lua#L36
4. check the response body 

```
--- response_body 
hello world
```

Reply:
Apologies for the delay. Thanks for the help with understanding how tests work.

I've added some information to the documentation which I think would have been enough for me to know how to setup this test myself. Let me know if it's acceptable.